allennlp_models/vision/dataset_readers/vgqa.py

```diff
@@ -0,0 +1,258 @@
+            # normalize self.images some more
+            # At this point, self.images maps filenames to full paths, but we want to map image ids to full paths.
+            filename_re = re.compile(r".*(\d{12})\.((jpg)|(png))")
+
+            def id_from_filename(filename: str) -> Optional[int]:
+                match = filename_re.fullmatch(filename)
+                if match is None:
+                    return None
+                return int(match.group(1))
+
+            if None in self.images:
+                del self.images[None]
+
+    @overrides
+    def _read(self, file_path: str):
+        # if the splits are using slicing syntax, honor it
+        question_slice, _ = utils.get_data_slice(file_path)
+
+        file_path = cached_path(file_path.split("[")[0], extract_archive=True)
+
+        logger.info("Reading file at %s", file_path)
+        questions = []
+        with open(file_path) as dataset_file:
+            dataset = json.load(dataset_file)
+        for data in dataset:
+            for qa in data["qas"]:
+                questions.append(qa)
+        questions = questions[question_slice]
+
+        question_dicts = list(self.shard_iterable(questions))
+        processed_images: Iterable[Optional[Tuple[Tensor, Tensor]]]
+        if self.produce_featurized_images:
+            # It would be much easier to just process one image at a time, but it's faster to process
+            # them in batches. So this code gathers up instances until it has enough to fill up a batch
+            # that needs processing, and then processes them all.
+
+            filenames = [f"{question_dict['image_id']}.jpg" for question_dict in question_dicts]
+            logger.info("images size: %s", len(self.images))
+            try:
+                processed_images = self._process_image_paths(
+                    self.images[filename] for filename in filenames
+                )
+            except KeyError as e:
+                print(self.images)
+                missing_id = e.args[0]
+                raise KeyError(
+                    missing_id,
+                    f"We could not find an image with the id {missing_id}. "
+                    "Because of the size of the image datasets, we don't download them automatically. "
+                    "Please go to https://visualqa.org/download.html, download the datasets you need, "
+                    "and set the image_dir parameter to point to your download location. This dataset "
+                    "reader does not care about the exact directory structure. It finds the images "
+                    "wherever they are.",
+                )
+        else:
+            processed_images = [None for _ in range(len(question_dicts))]
+
+        logger.info("Reading the dataset")
+        failed_instances_count = 0
+        attempted_instances_count = 0
+        for qa, processed_image in zip(question_dicts, processed_images):
+            question = qa["question"]
```

Comment:
I thought that Visual Genome gives you a single picture, and many questions for each. That's how it appears in the explorer at https://visualgenome.org/VGViz/explore?query=helping. But here we have one question per image. Is it already denormalized by the time we get here? Is the input data already denormalized?

Reply:
It's already been denormalized at this point. I separate the questions on lines 134-140.